daemon/exec: don't overwrite exit code if set

laurazard opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Related: https://github.com/moby/moby/issues/45795

If an exec fails to start, the code in this deferred block sets the exit code `126` (it wasn't possible to start the exec): https://github.com/moby/moby/blob/c7e42d855ec41c28702a1f6f6c2c28ecc7bd9424/daemon/exec.go#L181-L193

However, if we get far enough along attempting to start the exec, we also set the exit code from the error returned trying to start the exec: https://github.com/moby/moby/blob/c7e42d855ec41c28702a1f6f6c2c28ecc7bd9424/daemon/exec.go#L288-L291

For some situations (such as when the containers cmd/executable can't be found), the 2nd block returns the correct exit code (`127`) – https://github.com/moby/moby/blob/c7e42d855ec41c28702a1f6f6c2c28ecc7bd9424/daemon/errors.go#L164-L170

But then that error code is overwritten by 1st deferred block with `126`.

**- How I did it**

Change https://github.com/laurazard/moby/blob/c866a7e5f8caa882641cbba59102f457221314dc/daemon/exec.go#L181-L197 to in order to only set `126` if an error code has not already been set.

This change does not prevent the race between the deferred block locking/deleting the exec and setting the exit code race with `daemon.ProcessEvent`, if containerd publishes an exit event – I think that requires a more significant refactor of `daemon.ContainerExecStart` to split out the error handling for "pre-trying-to-start-the-exec" and "actually-started-the-exec". In the latter, we should set the error code explicitly and make sure to sync with `daemon.ProcessEvent`.

**- How to verify it**

```console
make TEST_FILTER=TestExecExitCode TEST_IGNORE_CGROUP_CHECK=1 test-integration
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->

**- A picture of a cute animal (not mandatory but encouraged)**

laurazard committed:
daemon/exec: don't overwrite exit code if set

If we fail to start an exec, the deferred error-handling block in [L181-L193](https://github.com/moby/moby/blob/c7e42d855ec41c28702a1f6f6c2c28ecc7bd9424/daemon/exec.go#L181-L193)
would set the exit code to `126` (`EACCES`). However, if we get far enough along
attempting to start the exec, we set the exit code according to the error returned
from starting the task [L288-L291](https://github.com/moby/moby/blob/c7e42d855ec41c28702a1f6f6c2c28ecc7bd9424/daemon/exec.go#L288-L291).

For some situations (such as `docker exec [some-container]
missing-binary`), the 2nd block returns the correct exit code (`127`)
but that then gets overwritten by the 1st block.

This commit changes that logic to only set the default exit code `126`
if the exit code has not been set yet.

Signed-off-by: Laura Brehm <laurabrehm@hey.com>

laurazard commented:
Turns out we'd baked some expectations of [this into tests](https://github.com/moby/moby/actions/runs/11052960324/job/30706287545?pr=48552#step:7:97):
```console
tests/integration/models_containers_test.py:363: in test_exec_run_failed
    assert exec_output[0] == 126
E   assert 127 == 126
```

https://github.com/docker/docker-py/blob/a3652028b1ead708bd9191efb286f909ba6c2a49/tests/integration/models_containers_test.py#L356-L363

Which leads me to wonder if we can reasonably fix this without breaking a lot of users. I want to believe that it's quite a specific edge case (`126`vs`127` when the executable can't be found), but 😨

thaJeztah commented:
I need to dig if I can find discussion related to it, but ISTR there also was some convention to make "exit code + 125 (or something) to distinguish "client" from "container" errors, but I keep forgetting.

For the docker-py test; I see it was added in this PR; no specific mention of that exit code, so indeed likely written based on observed behavior;

- https://github.com/docker/docker-py/pull/1872

laurazard commented:
I remember something about that, and I know I fixed some error causing exit codes to overflow/underflow in some instances related to that at some point 😅 

Regardless, `127` would be the normal exit code for "not found" here, and permission denied would generally by a `125+1` I guess (`1` being what one would normally get trying to execute a binary you have no permission to).

laurazard committed:
tests: skip docker-py exec exit code test

Temporarily skip the exec run failed exit code test in `docker-py` –
https://github.com/docker/docker-py/blob/a3652028b1ead708bd9191efb286f909ba6c2a49/tests/integration/models_containers_test.py#L356-L363

We can reenable this after the PR fixing the expected exit code in that
test is merged/released/included – https://github.com/docker/docker-py/pull/3290

Signed-off-by: Laura Brehm <laurabrehm@hey.com>

thaJeztah commented on daemon/exec.go:
```diff
@@ -183,8 +183,10 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, optio
-			exitCode := 126
-			ec.ExitCode = &exitCode
+			if ec.ExitCode == nil {
+				exitCode := 126
```
We can probably use the consts here to give  more context what this magic number is; not sure if there's other logic in that file as well that could be useful here (just stepped away from my computer 🥹); https://github.com/moby/moby/blob/62120a727c7e28609fec296f23c59dda62078b8a/daemon/errors.go#L122-L125

laurazard replied:
I'd like to refactor a lot of this code, but was still trying to look at whether it made sense to do that in this PR or a follow-up. I also wanted to see if there was another small change I could do here to address this issue in particular without possibly breaking a lot of other logic.

laurazard replied:
The more I look at this, the more I dislike that `defer` error handling block. We capture a bunch of errors and default to `126` (`EACCES`) for any reason that we fail to start the exec. It's hard to have confidence changing anything here since so many different things could cause the `126` to be set, including failing to unmarshal a json spec, etc.

thaJeztah reviewed (approved):
LGTM

corhere commented:
> I need to dig if I can find discussion related to it, but ISTR there also was some convention to make "exit code + 125 (or something) to distinguish "client" from "container" errors, but I keep forgetting.

I think the convention comes from the shell?
https://www.gnu.org/software/bash/manual/html_node/Exit-Status.html
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_08_02

Not found -> 126
Not executable -> 127
Killed by signal N -> 128+N

corhere reviewed (commented):
Does this PR also fix the inconsistency with the `exitCode` reported in the `exec_die` daemon event, i.e. fully fix #45795? If so, I would like to see the regression test cover both the Engine API response and the daemon event. If not, I would appreciate having a brief mention in the PR description of why that aspect of the defect is not as straightforward to fix.

laurazard commented:
Oh, I meant to update the PR description to include a mention of that. I think preventing the race between that deferred block locking/deleting the exec and setting the exit code race with `daemon.ProcessEvent` requires a more significant refactor of `daemon.ContainerExecStart` – ultimately, this code is a bit of a mess, we default to `126` for **way too many** paths, and there are there are two distinct parts there – one for which we want the "catch-all set 126 error handling", and a point after which we should handle the exit differently, relying on the exit code reported, etc.

laurazard commented:
I updated the PR description, thanks @corhere.
